### ISSUE TITLE
conformity_custom_rule: boolean values will be converted in a string value and the custom rule will fail

### ISSUE THREAD

**Issue by ari-neto** [trendmicro/terraform-provider-conformity#72]
# conformity_custom_rule: boolean values will be converted in a string value and the custom rule will fail

**Resource:** custom rule
**Documentation:** [https://registry.terraform.io/providers/trendmicro/conformity/latest/docs/resources/conformity_custom_rule](https://registry.terraform.io/providers/trendmicro/conformity/latest/docs/resources/conformity_custom_rule)

## Problem description

Reserved values as "null", "true", "false", are considered string by the provider. A "Null" cannot be used as a value using the provider, although it can be used as value on the dry-run endpoint, the provider will convert the boolean in a string, and it will affect the rule behavior.

In the example bellow when we defined the value as true, the provider convert in a string value "true" when create the resource on the Conformity platform,  the rule will have a different behavior than the intended, and the buckets that do not have a policy defined to allow a cloudfront access will fail, because the boolean operator "isNullOrUndefined" will not be executed correctally on the backend. 

**Provider version tested:** v0.5.4 and v0.5.6

## Detailed description

Conditions defined in the dry-run endpoint (custom-rules/run):

~~~json
"conditions": {
            "any": [
                {
                    "value": true,
                    "operator": "isNullOrUndefined",
                    "fact": "cloudFrontInPolicy"
                },
                {
                    "value": ".*\"Service\".?\\S\"cloudfront.amazonaws.com\".*",
                    "operator": "pattern",
                    "fact": "cloudFrontInPolicy"
                }
            ]
}
~~~

HCL Code:

~~~hcl
resource "conformity_custom_rule" "custom_rule_s3_cf_001"{
  name                = "S3 bucket access restricted through CloudFront or without bucket policy configuration"
  description         = "This rule checks if CloudFront is restricting S3 bucket access configured by Cloud Security Team"
    remediation_notes = "If this is broken, please follow these steps:\n1. https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-restricting-access-to-s3.html \n2. https://aws.amazon.com/blogs/networking-and-content-delivery/amazon-cloudfront-introduces-origin-access-control-oac/\n updated"
    service           = "S3"
    resource_type     = "s3-bucket"
    categories        = ["security"]
    severity          = "MEDIUM"
    cloud_provider    = "aws"
    enabled           = false
    attributes {
        name     = "cloudFrontInPolicy"
        path     = "data.Policy"
        required = true
      }
      rules {
        operation = "any"
        conditions {
          fact     = "cloudFrontInPolicy"
          operator = "equal"
          value    =  null
        }

        conditions {
          fact     = "cloudFrontInPolicy"
          operator = "pattern"
          value    =  ".*\"Service\".?\\S\"cloudfront.amazonaws.com\".*"
        }
        event_type = "Bucket is restricted by CloudFront or does not have access policy attached"
      }
  }
~~~

Terraform shows that the bolean value could be transformed in a string value:

~~~json
+ rules {
    + event_type = "Bucket is restricted by CloudFront or does not have access policy attached"
    + operation  = "any"

    + conditions {
        + fact     = "cloudFrontInPolicy"
        + operator = "isNullOrUndefined"
        + value    = "true"
    }
    + conditions {
        + fact     = "cloudFrontInPolicy"
        + operator = "pattern"
        + value    = ".*\"Service\".?\\S\"cloudfront.amazonaws.com\".*"
    }
}
~~~

And when we see the rule that is actually created by the terraform code:

~~~json
"conditions": {
    "any": [
        {
            "value": "true",
            "operator": "isNullOrUndefined",
            "fact": "cloudFrontInPolicy"
        },
        {
            "value": ".*\"Service\".?\\S\"cloudfront.amazonaws.com\".*",
            "operator": "pattern",
            "fact": "cloudFrontInPolicy"
        }
    ]
}
~~~

And differently the result we have using the dry-run endpoint a bucket that does not have a policy defined will fail. And this configuration is not supported in the dry-run endpoint, if we try to define the boolean as a string, the dry-run endpoint return this error:

~~~json
{
    "errors": [
        {
            "status": 422,
            "source": {
                "pointer": "/data/attributes/configuration/rules"
            },
            "detail": "Value of isNullOrUndefined operator should be boolean"
        }
    ]
}
~~~

**Comment by johan1252** [comment]
Thanks @ari-neto, looks like only string is supported at the moment. We'll have a look at supporting the other value types for future.

**Comment by kim-cloudconformity** [comment]
Hi @ari-neto , this has been patched in v0.5.7. To support values other than string, we will use the Terraform `jsonencode()` function in the Terraform template which will then be stored in Conformity with the intended data type. An example can be found below, but more information has been updated in the docs [here](https://registry.terraform.io/providers/trendmicro/conformity/latest/docs/resources/conformity_custom_rule).

e.g. For a custom rule where the value is a boolean data type
```hcl
conditions {
  fact = "cloudFrontInPolicy"
  operator = "isNullOrUndefined"
  value = jsonencode(true)
}
```
